app/scripts/modules/core/src/domain/IManagedEntity.ts

```diff
@@ -103,10 +103,35 @@ export interface IManagedArtifactVersion {
     statelessConstraints?: IStatelessConstraint[];
   }>;
   build?: {
-    id: number;
+    id: number; // deprecated, use number
+    number: string;
+    uid: string;
```

Comment:
yeah, I could get a little more aggressive about making these explicitly optional. I got away with it because in our installation/setup of TS all values are implicitly union'd with `null`, so technically _everything_ is actually nullable all the time everywhere.

I would like to understand the behavior of these fields better in terms of when they will/won't show up — the way I was thinking about them is that you hit a couple of different states:

1. Build/git data couldn't be fetched, all this stuff is totally missing
2. Data could be fetched, only a couple things are _really_ nullable once you know that (i.e. the pull request data might be missing because it wasn't part of a PR, `completedAt` may not be there if the build isn't done yet)

Is that accurate? or am I expected to be defensive about _every single field_ on these objects independent of the others and whether they exist, not just the ones i mentioned (prs, `completedAt`)? Right now the way the code is written nothing will break either way, but as we use this data more having each of the field independently nullable will get extremely complicated to code against.

(cc @gal-yardeni)

Reply:
@erikmunson let me pull all the possible options -- will link them here shortly :)

Reply:
Here is an example which contians all the possible responses from `/application/gyardeniapp?entities=artifacts` endpoint:
- build 135 is currently running
- build 134 is finished
- build 133 has only the default information
```
// https://keel-prestaging.us-west-2.spinnaker.mgmt.netflix.net/application/gyardeniapp?entities=artifacts,resources

{
  "applicationPaused": false,
  "hasManagedResources": true,
  "currentEnvironmentConstraints": [
    
  ],
  "artifacts": [
    {
      "name": "gyardeni-app-server",
      "type": "deb",
      "reference": "gyardeni-deb",
      "versions": [
        {
          "version": "gyardeni-app-server-0.0.1~snapshot.202009181544-h135.fcb0ca1",
          "displayName": "0.0.1~snapshot.202009181544",
          "environments": [
            {
              "name": "testing",
              "state": "pending"
            }
          ],
          "build": {
            "id": 135,
            "number": "135",
            "uid": "7da37eb6-298e-4472-b110-d59d0c339580",
            "job": {
              "link": "https://buildstest.builds.test.netflix.net/job/users-gyardeni-gyardeni-app-build/135/",
              "name": "users-gyardeni-gyardeni-app-build"
            },
            "startedAt": "2020-09-18T15:43:56.710Z",
            "status": "BUILDING"
          },
          "git": {
            "commit": "fcb0ca1",
            "author": "gyardeni",
            "project": "~gyardeni",
            "branch": "master",
            "repo": {
              "name": "gyardeni-app",
              "link": ""
            },
            "pullRequest": {
              "number": "",
              "url": ""
            },
            "commitInfo": {
              "sha": "fcb0ca1a04bb835b6283b7e7cfaa3629c0661f14",
              "link": "https://stash.corp.netflix.com/projects/~gyardeni/repos/gyardeni-app/commits/fcb0ca1a04bb835b6283b7e7cfaa3629c0661f14",
              "message": "Update README.md"
            }
          }
        },
        {
          "version": "gyardeni-app-server-0.0.1~snapshot.202009172025-h134.fcb0ca1",
          "displayName": "0.0.1~snapshot.202009172025",
          "environments": [
            {
              "name": "testing",
              "state": "current",
              "deployedAt": "2020-09-17T20:41:24.449Z"
            }
          ],
          "build": {
            "id": 134,
            "number": "134",
            "uid": "1ef0915a-65b4-4bd8-893f-ee9f6692612d",
            "job": {
              "link": "https://buildstest.builds.test.netflix.net/job/users-gyardeni-gyardeni-app-build/134/",
              "name": "users-gyardeni-gyardeni-app-build"
            },
            "startedAt": "2020-09-17T20:25:10.995Z",
            "status": "BUILDING"
          },
          "git": {
            "commit": "fcb0ca1",
            "author": "gyardeni",
            "project": "~gyardeni",
            "branch": "master",
            "repo": {
              "name": "gyardeni-app",
              "link": ""
            },
            "pullRequest": {
              "number": "",
              "url": ""
            },
            "commitInfo": {
              "sha": "fcb0ca1a04bb835b6283b7e7cfaa3629c0661f14",
              "link": "https://stash.corp.netflix.com/projects/~gyardeni/repos/gyardeni-app/commits/fcb0ca1a04bb835b6283b7e7cfaa3629c0661f14",
              "message": "Update README.md"
            }
          }
        },
        {
          "version": "gyardeni-app-server-0.0.1~snapshot.202009171824-h133.fcb0ca1",
          "displayName": "0.0.1~snapshot.202009171824",
          "environments": [
            {
              "name": "testing",
              "state": "previous",
              "deployedAt": "2020-09-17T18:38:43.753Z",
              "replacedAt": "2020-09-17T20:41:24.452Z",
              "replacedBy": "gyardeni-app-server-0.0.1~snapshot.202009172025-h134.fcb0ca1"
            }
          ],
          "build": {
            "id": 133
          },
          "git": {
            "commit": "fcb0ca1"
          }
        },
        {
          "version": "gyardeni-app-server-0.0.1~snapshot.202009171231-h132.fcb0ca1",
          "displayName": "0.0.1~snapshot.202009171231",
          "environments": [
            {
              "name": "testing",
              "state": "previous",
              "deployedAt": "2020-09-17T12:48:57.003Z",
              "replacedAt": "2020-09-17T18:38:43.755Z",
              "replacedBy": "gyardeni-app-server-0.0.1~snapshot.202009171824-h133.fcb0ca1"
            }
          ],
          "build": {
            "id": 132
          },
          "git": {
            "commit": "fcb0ca1"
          }
        },
        }

Reply:
👍 thanks gal — seems like my understanding of the two states (either you have full data or you don't have any data) is roughly accurate, which is useful.